features/ForceExpiration.php

```diff
@@ -86,23 +82,67 @@ private function _addPasswordExpirationCheck() : void {
 	/**
 	 * Retrieve the password expiration date
 	 * @param User $user
-	 * @return DateTime
+	 * @return DateTime Most recent time of password change
 	 */
-	private function _getExpirationDate(User $user) : DateTime {
-		$expirationDate = new DateTime($user->getData("{$this->_plugin->getSettingsName()}::lastPasswordUpdate") ?: $user->getDateRegistered());
-		$expirationDate->modify("{$this->_passwordLifetime} day");
-		return $expirationDate;
+	private function _getExpirationDate(\PKP\user\User $user) : \DateTime {
+		/* @var $storedPasswordsDao StoredPasswordsDAO */
+		$storedPasswordsDao = DAORegistry::getDAO('StoredPasswordsDAO');
+		$storedPasswords = $storedPasswordsDao->newDataObjects($user->getId());
+		$mostRecent = new \DateTime ($storedPasswordsDao->getMostRecent($storedPasswords));
+		if (!$mostRecent) {
+			$mostRecent = new \DateTime ($user->getDateRegistered());
+		}
+
+		$mostRecent->modify("{$this->_passwordLifetime} day");
+		return $mostRecent;
 	}
 
-	private function _handleNotification(User $user) : void {
+	/**
+	 * Updates the time of the users last password change
+	 * @param string $hook The hook name
+	 * @param array $args Arguments of the hook
+	 * @return bool false if process completes
+	 */
+	public function rememberPasswordDate($hook, $args) {
+		if ($hook == 'changepasswordform::execute' && get_class($args[0]) == 'PKP\user\form\ChangePasswordForm') {
+			$user = $args[0]->getUser();
+		}
+		elseif ($hook == 'loginchangepasswordform::execute' && get_class($args[0]) == 'PKP\user\form\LoginChangePasswordForm') {
+			$user = Repo::user()->getByUsername($args[0]->getData('username'), false);
+		}
+		else {
+		}
+
+		$storedPasswordsDao = DAORegistry::getDAO('StoredPasswordsDAO');
+		if ($user) {
+			$storedPasswords = $storedPasswordsDao->getByUserId($user->getId());
```

Comment:
Is there a race condition between this feature's recording of the last changed date and the reuse feature's recording of the password itself?

Reply:
Would removing the action to update the lastChangeTime in limit reuse solve the race condition?

Reply:
Have you identified if there is, in fact, a race condition?  If so, what is the nature of the race condition?  What would happen if each feature was enabled independently, mutually, and sequentially?